jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/BlossomVInitializer.java

```diff
@@ -0,0 +1,317 @@
+     * time determine its generic counterpart.
+     */
+    private List<E> graphEdges;
+
+    /**
+     * Creates a new BlossomVInitializer instance
+     *
+     * @param graph the graph to search matching in
+     */
+    public BlossomVInitializer(Graph<V, E> graph) {
+        this.graph = graph;
+    }
+
+    /**
+     * Converts the generic graph representation into the data structure form convenient for the algorithm
+     * and initializes the matching according to the strategy specified in {@code options}
+     *
+     * @param options the options of the algorithm
+     * @return the state object with all necessary for the algorithm information
+     */
+    public BlossomVState<V, E> initialize(BlossomVOptions options) {
+        initGraph();
+        int treeNum;
+        BlossomVOptions.InitializationType type = options.initializationType;
+        if (type == GREEDY) {
+            treeNum = initGreedy();
+        } else {
+            // simple initialization
+            treeNum = nodeNum;
+            for (BlossomVNode node : nodes) {
+                node.isOuter = true;
+            }
+        }
+        allocateTrees();
+        // initializing tree edges and adding cross-tree edges to corresponding heaps
+        initAuxiliaryGraph();
+        return new BlossomVState<>(graph, nodes, edges, nodeNum, edgeNum, treeNum, graphVertices, graphEdges, options);
+    }
+
+    /**
+     * Helper method to convert the generic graph representation into the form convenient for the algorithm
+     */
+    private void initGraph() {
+        nodeNum = graph.vertexSet().size();
+        int expectedEdgeNum = graph.edgeSet().size();
+        nodes = new BlossomVNode[nodeNum + 1];
+        edges = new BlossomVEdge[expectedEdgeNum];
+        graphVertices = new ArrayList<>(nodeNum);
+        graphEdges = new ArrayList<>(expectedEdgeNum);
+        HashMap<V, BlossomVNode> vertexMap = new HashMap<>(nodeNum);
+        int i = 0;
+        // mapping nodes
+        for (V vertex : graph.vertexSet()) {
+            nodes[i] = new BlossomVNode(i);
+            graphVertices.add(vertex);
+            vertexMap.put(vertex, nodes[i]);
+            i++;
+        }
+        nodes[nodeNum] = new BlossomVNode(nodeNum);  // auxiliary node to keep track of the first item in the linked list of tree roots
+        i = 0;
+        // mapping edges
+        for (E e : graph.edgeSet()) {
+            BlossomVNode source = vertexMap.get(graph.getEdgeSource(e));
+            BlossomVNode target = vertexMap.get(graph.getEdgeTarget(e));
+            if (source != target) { // we avoid self-loops in order to support pseudographs
+                edgeNum++;
+                BlossomVEdge edge = addEdge(source, target, graph.getEdgeWeight(e), i);
+                edges[i] = edge;
+                graphEdges.add(e);
+                i++;
+            }
+        }
+    }
```

Comment:
Initialization is a little convoluted. I tried restructuring it, but then gave up :)

Reply:
I've separated simple and greedy initialization into their own methods. Now it should be cleaner.